File: xml-sdk/src/test/java/com/cleanster/xml/PaymentMethodsTest.java
package com.cleanster.xml;

import com.cleanster.xml.api.PaymentMethodsXmlApi;
import com.cleanster.xml.client.XmlConverter;
import com.cleanster.xml.client.XmlHttpClient;
import com.cleanster.xml.model.PaymentMethod;
import com.cleanster.xml.model.XmlApiResponse;
import org.junit.jupiter.api.*;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

class PaymentMethodsTest {

    private XmlHttpClient        http;
    private PaymentMethodsXmlApi api;

    private static final String PM_JSON   = "{\"success\":true,\"message\":\"OK\","
            + "\"data\":{\"id\":10,\"brand\":\"Visa\",\"last4\":\"4242\"}}";
    private static final String LIST_JSON = "{\"success\":true,\"message\":\"OK\","
            + "\"data\":[{\"id\":10,\"brand\":\"Visa\",\"last4\":\"4242\"}]}";
    private static final String OK_JSON   = "{\"success\":true,\"message\":\"OK\",\"data\":{}}";

    @BeforeEach void setUp() {
        XmlHttpClient real = XmlHttpClient.builder().baseUrl("http://dummy").accessKey("key").build();
        http = spy(real);
        api  = new PaymentMethodsXmlApi(http);
    }

    // ── getSetupIntentDetails ──────────────────────────────────────────────────

    @Test void getSetupIntentDetails_callsGet() {
        doReturn(OK_JSON).when(http).get("/v1/payment-methods/setup-intent-details");
        api.getSetupIntentDetails();
        verify(http).get("/v1/payment-methods/setup-intent-details");
    }

    @Test void getSetupIntentDetails_returnsSuccess() {
        doReturn(OK_JSON).when(http).get("/v1/payment-methods/setup-intent-details");
        assertTrue(api.getSetupIntentDetails().isSuccess());
    }

    // ── getPaypalClientToken ───────────────────────────────────────────────────

    @Test void getPaypalClientToken_callsGet() {
        doReturn(OK_JSON).when(http).get("/v1/payment-methods/paypal-client-token");
        api.getPaypalClientToken();
        verify(http).get("/v1/payment-methods/paypal-client-token");
    }

    @Test void getPaypalClientToken_returnsSuccess() {
        doReturn(OK_JSON).when(http).get("/v1/payment-methods/paypal-client-token");
        assertTrue(api.getPaypalClientToken().isSuccess());
    }

    // ── addPaymentMethod ───────────────────────────────────────────────────────

    @Test void addPaymentMethod_callsPost() {
        doReturn(PM_JSON).when(http).post(eq("/v1/payment-methods"), any());
        api.addPaymentMethod(Map.of("paymentMethodId", "pm_test_123"));
        verify(http).post(eq("/v1/payment-methods"), any());
    }

    @Test void addPaymentMethod_returnsParsedPaymentMethod() {
        doReturn(PM_JSON).when(http).post(eq("/v1/payment-methods"), any());
        XmlApiResponse<PaymentMethod> resp = api.addPaymentMethod(Map.of("paymentMethodId", "pm_test"));
        assertTrue(resp.isSuccess());
        assertEquals(10, (int) resp.getData().getId());
    }

    // ── listPaymentMethods ─────────────────────────────────────────────────────

    @Test void listPaymentMethods_callsGet() {
        doReturn(LIST_JSON).when(http).get("/v1/payment-methods");
        api.listPaymentMethods();
        verify(http).get("/v1/payment-methods");
    }

    @Test void listPaymentMethods_returnsList() {
        doReturn(LIST_JSON).when(http).get("/v1/payment-methods");
        XmlApiResponse<List<PaymentMethod>> resp = api.listPaymentMethods();
        assertTrue(resp.isSuccess());
        assertEquals(1, resp.getData().size());
        assertEquals("Visa", resp.getData().get(0).getBrand());
    }

    // ── deletePaymentMethod ────────────────────────────────────────────────────

    @Test void deletePaymentMethod_callsDeleteWithId() {
        doReturn(OK_JSON).when(http).delete("/v1/payment-methods/10");
        api.deletePaymentMethod(10);
        verify(http).delete("/v1/payment-methods/10");
    }

    @Test void deletePaymentMethod_returnsSuccess() {
        doReturn(OK_JSON).when(http).delete("/v1/payment-methods/10");
        assertTrue(api.deletePaymentMethod(10).isSuccess());
    }

    // ── setDefaultPaymentMethod ────────────────────────────────────────────────

    @Test void setDefaultPaymentMethod_callsPutWithId() {
        doReturn(OK_JSON).when(http).put(eq("/v1/payment-methods/10/default"), any());
        api.setDefaultPaymentMethod(10);
        verify(http).put(eq("/v1/payment-methods/10/default"), any());
    }

    @Test void setDefaultPaymentMethod_returnsSuccess() {
        doReturn(OK_JSON).when(http).put(eq("/v1/payment-methods/10/default"), any());
        assertTrue(api.setDefaultPaymentMethod(10).isSuccess());
    }

    // ── JAXB ───────────────────────────────────────────────────────────────────

    @Test void paymentMethod_toXml_isValidXml() {
        PaymentMethod pm = new PaymentMethod(); pm.setId(1); pm.setBrand("Visa"); pm.setLast4("4242");
        assertTrue(XmlConverter.isXml(XmlConverter.toXml(pm)));
    }

    @Test void paymentMethod_fromXml_roundTrip() {
        PaymentMethod o = new PaymentMethod(); o.setId(5); o.setBrand("MC"); o.setLast4("5678");
        PaymentMethod r = XmlConverter.fromXml(XmlConverter.toXml(o), PaymentMethod.class);
        assertEquals(5, (int) r.getId());
        assertEquals("MC", r.getBrand());
        assertEquals("5678", r.getLast4());
    }
}

File: kotlin-sdk/src/test/kotlin/com/cleanster/PaymentMethodsTest.kt
package com.cleanster

import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class PaymentMethodsTest {

    private lateinit var mock: MockHttpEngine
    private lateinit var client: CleansterClient

    @BeforeEach fun setUp() {
        mock   = MockHttpEngine()
        client = testClient(mock)
    }

    @Test fun `getSetupIntentDetails sends GET`() = runTest {
        mock.succeed(mapOf("clientSecret" to "seti_xxx"))
        client.paymentMethods.getSetupIntentDetails()
        assertEquals("GET", mock.capturedMethod)
    }

    @Test fun `getSetupIntentDetails correct path`() = runTest {
        mock.succeed(mapOf("clientSecret" to "seti_xxx"))
        client.paymentMethods.getSetupIntentDetails()
        assertTrue(mock.capturedUrl?.endsWith("/v1/payment-methods/setup-intent-details") == true)
    }

    @Test fun `getPayPalClientToken sends GET`() = runTest {
        mock.succeed(mapOf("clientToken" to "paypal_xxx"))
        client.paymentMethods.getPayPalClientToken()
        assertEquals("GET", mock.capturedMethod)
    }

    @Test fun `getPayPalClientToken correct path`() = runTest {
        mock.succeed(mapOf("clientToken" to "paypal_xxx"))
        client.paymentMethods.getPayPalClientToken()
        assertTrue(mock.capturedUrl?.endsWith("/v1/payment-methods/paypal-client-token") == true)
    }

    @Test fun `addPaymentMethod sends POST`() = runTest {
        mock.succeed(mapOf("id" to 55.0, "type" to "card"))
        client.paymentMethods.addPaymentMethod("pm_abc")
        assertEquals("POST", mock.capturedMethod)
    }

    @Test fun `addPaymentMethod correct path`() = runTest {
        mock.succeed(mapOf("id" to 55.0))
        client.paymentMethods.addPaymentMethod("pm_abc")
        assertTrue(mock.capturedUrl?.endsWith("/v1/payment-methods") == true)
    }

    @Test fun `addPaymentMethod encodes token`() = runTest {
        mock.succeed(mapOf("id" to 55.0))
        client.paymentMethods.addPaymentMethod("pm_abc123")
        assertEquals("pm_abc123", mock.capturedBody?.get("paymentMethodId"))
    }

    @Test fun `addPaymentMethod decodes id`() = runTest {
        mock.succeed(mapOf("id" to 55.0, "type" to "card", "brand" to "visa"))
        val resp = client.paymentMethods.addPaymentMethod("pm_xxx")
        assertEquals(55, resp.data?.id)
    }

    @Test fun `addPaymentMethod decodes type`() = runTest {
        mock.succeed(mapOf("id" to 55.0, "type" to "paypal"))
        val resp = client.paymentMethods.addPaymentMethod("nonce_xxx")
        assertEquals("paypal", resp.data?.type)
    }

    @Test fun `getPaymentMethods sends GET`() = runTest {
        mock.succeedList(); client.paymentMethods.getPaymentMethods()
        assertEquals("GET", mock.capturedMethod)
    }

    @Test fun `getPaymentMethods correct path`() = runTest {
        mock.succeedList(); client.paymentMethods.getPaymentMethods()
        assertTrue(mock.capturedUrl?.endsWith("/v1/payment-methods") == true)
    }

    @Test fun `setDefaultPaymentMethod sends PUT`() = runTest {
        mock.succeedEmpty(); client.paymentMethods.setDefaultPaymentMethod(55)
        assertEquals("PUT", mock.capturedMethod)
    }

    @Test fun `setDefaultPaymentMethod correct path`() = runTest {
        mock.succeedEmpty(); client.paymentMethods.setDefaultPaymentMethod(55)
        assertTrue(mock.capturedUrl?.endsWith("/v1/payment-methods/55/default") == true)
    }

    @Test fun `deletePaymentMethod sends DELETE`() = runTest {
        mock.succeedEmpty(); client.paymentMethods.deletePaymentMethod(55)
        assertEquals("DELETE", mock.capturedMethod)
    }

    @Test fun `deletePaymentMethod correct path`() = runTest {
        mock.succeedEmpty(); client.paymentMethods.deletePaymentMethod(55)
        assertTrue(mock.capturedUrl?.endsWith("/v1/payment-methods/55") == true)
    }
}
